File: bench/bench.go
package main

import (
	_ "embed" // 1. Import embed (the underscore is required)
	"log"
	"net/http"
	"time"

	"github.com/starfederation/datastar-go/datastar"
)

func main() {
	http.HandleFunc("/", handler)
	http.HandleFunc("/log", handlerLogged)
	http.HandleFunc("/sse", sseHandler)

	log.Println("Go Datastar SSE Server running at http://localhost:8091")
	if err := http.ListenAndServe(":8091", nil); err != nil {
		log.Fatal(err)
	}
}

//go:embed index.html
var indexHTML string

//go:embed sse.html
var sseHTML string

func handler(w http.ResponseWriter, r *http.Request) {
	w.Write([]byte(indexHTML))
}

func handlerLogged(w http.ResponseWriter, r *http.Request) {
	t1 := time.Now().UnixMicro()
	w.Write([]byte(indexHTML))
	println("Go index handler took", time.Now().UnixMicro()-t1, "microseconds")
}

func sseHandler(w http.ResponseWriter, r *http.Request) {
	t1 := time.Now().UnixMicro()
	sse := datastar.NewSSE(w, r)
	sse.PatchElements(indexHTML)
	println("Go index handler took", time.Now().UnixMicro()-t1, "microseconds")
}

File: bench/bench.zig
const std = @import("std");
const httpz = @import("httpz");
const datastar = @import("datastar");
const builtin = @import("builtin");

const Allocator = std.mem.Allocator;
var debug_allocator: std.heap.DebugAllocator(.{}) = .init;

pub fn main() !void {
    const gpa, const is_debug = gpa: {
        break :gpa switch (builtin.mode) {
            .Debug, .ReleaseSafe => .{ debug_allocator.allocator(), true },
            .ReleaseFast, .ReleaseSmall => .{ std.heap.smp_allocator, false },
        };
    };
    defer if (is_debug) {
        _ = debug_allocator.deinit();
    };

    var server = try httpz.Server(void).init(gpa, .{
        .address = "0.0.0.0",
        .port = 8098,
    }, {});
    var router = try server.router(.{});
    router.get("/", handler, .{});
    router.get("/sse", sseHandler, .{});
    std.debug.print("Zig Datastar SSE Server running at http://localhost:8098\n", .{});
    return server.listen();
}

pub fn handler(_: *httpz.Request, res: *httpz.Response) !void {
    const t1 = std.time.microTimestamp();
    res.body = @embedFile("index.html");
    std.debug.print("Zig index handler took {} microseconds\n", .{std.time.microTimestamp() - t1});
}

pub fn sseHandler(req: *httpz.Request, res: *httpz.Response) !void {
    const t1 = std.time.microTimestamp();
    var sse = try datastar.NewSSE(req, res);

    try sse.patchElements(@embedFile("sse.html"), .{});

    res.body = sse.buffered();
    std.debug.print("Zig SSE handler took {} microseconds\n", .{std.time.microTimestamp() - t1});
}
